pkgs/development/interpreters/spidermonkey/common.nix

```
@@ -0,0 +1,179 @@
+{ version, hash }:
+
+{ callPackage
+, lib
+, stdenv
+, fetchurl
+, fetchpatch
+
+# build time
+, buildPackages
+, cargo
+, m4
+, perl
+, pkg-config
+, python39
+, rustc
+, which
+, zip
+, autoconf213
+, yasm
+, xcbuild
+
+# runtime
+, icu
+, nspr
+, readline
+, zlib
+, libobjc
+, libiconv
+}:
+
+stdenv.mkDerivation (finalAttrs: rec {
+  pname = "spidermonkey";
+  inherit version;
+
+  src = fetchurl {
+    url = "mirror://mozilla/firefox/releases/${version}esr/source/firefox-${version}esr.source.tar.xz";
+    inherit hash;
+  };
+
+  outputs = [ "out" "dev" ];
+  setOutputFlags = false; # Configure script only understands --includedir
+
+  nativeBuildInputs = [
+    cargo
+    m4
+    perl
+    pkg-config
+    python39 # 78 requires python up to 3.9
```

Avoid introducing extra Python dependency into the build closure on systems that use new Spidermonkey just because the old one does not build with it:

```suggestion
    # 78 requires python up to 3.9
    (if lib.versionOlder version "91" then python3 else python39)
```